magma/types/ready_valid.py

```diff
@@ -245,6 +242,25 @@ def no_enq(self, when=True):
             self.data @= 0
 
 
+class ReadyValidMonitorKind(ReadyValidKind):
+    def __getitem__(cls, T: Union[Type, MagmaProtocol]):
+        fields = {"valid": In(Bit), "ready": In(Bit)}
+        _maybe_add_data(cls, fields, T, In)
+        return type(f"{cls}[{T}]", (cls, ), fields)
+
+    def flip(cls):
```

Comment:
should `qualify()` also raise an error here? should that be the case for all of the derived types? e.g. `m.Out(m.In(m.Bit))` is valid but `m.Out(m.In(m.ReadyValid[...]))` should be disallowed?

Reply:
This inherits the qualify method from ReadyValidKind, which only allows calling In on a readyvalid type.  so this works
```
    T = m.Consumer(m.ReadyValid[m.Bits[8]])
    T = m.In(T)
    T = m.In(T)
```
but this doesn't work
```
    m.Out(T)
```
and raises:

```
    def qualify(cls, direction):
        if direction is not Direction.In:
>           raise TypeError(f"Cannot qualify ReadyValid with {direction}")
E           TypeError: Cannot qualify ReadyValid with Direction.Out

magma/types/ready_valid.py:50: TypeError
```

So right now qualify does raise an error if you use anything other than m.In

Reply:
In general, I think m.Out(m.In(m.Bit)) makes sense because you may want to get a specific version of a generic type (e.g. take any directed type and turn it into an Out).  However, for readyvalid, having an Out of readyvalid type doesn't really make sense, so we override the default behavior to raise an error because in this case the semantics are not well formed (we can monitor a ReadyValid as a In, but it doesn't make sense really to drive both the In and the Out port, well maybe it would for some mock driver for testing, but unless this use case pops up I say we don't support it for now).